### ISSUE TITLE
Understanding the post-processing

### ISSUE THREAD

**Issue by MickaMickaMicka** [hirotomusiker/CLRerNet#58]
Thank you for your great work, the paper and repository are impressive!

We are currently trying to get an ONNX variant (and a smaller model) running in an isolated environment (without your repository).
We converted successfully to ONNX and we are getting 4 output layers with correct output dimensions with that ONNX model:

logits: 1x192x2
anchors: 1x192x3
lengths: 1x192x1
xs: 1x192x72

However we are struggeling a bit with the post-processing.

The logits look like they make sense, giving softmax values beween 0.2 and 0.99 and the raw values are typically +a for the first value and about -a for the second value.

We are scaling the anchor x and y by the image size, which looks like reasonable results (except some few anchors outside of the image and at the very top of the image). 

However the theta angles are strange. The minimum of the (raw dnn output) thetas is 0.067509 and maximum is 1.38438 for a test image. Which looks like the model outputs positive values between 0 and pi/2? But to cover both perspective directions, a range of 180° would be necessary (0 to pi or -pi/2 to pi/2). Which post-processing is necessary on the theta angles to get the right image line angle?

Probably more questions will rise for how to apply the xs values...

Is the post-processing explained somewhere ? The comments in the code aren't sufficient and we are having a hard time to trace down the code because of mmdet and other interoperations.

**Comment by hirotomusiker** [comment]
Thank you for your questions.

### Post-processing pipeline

https://github.com/hirotomusiker/CLRerNet/blob/main/libs/models/dense_heads/clrernet_head.py#L397-L518

1. Lanes with higher confidence values (softmax(logits)) than threshold are selected. [[code](https://github.com/hirotomusiker/CLRerNet/blob/main/libs/models/dense_heads/clrernet_head.py#L414-L427)]
2. [NMS](https://github.com/hirotomusiker/CLRerNet/blob/main/libs/models/layers/nms/src/nms_kernel.cu) is applied to remove overlapped lanes. It's not the common bounding-box NMS. [[code](https://github.com/hirotomusiker/CLRerNet/blob/main/libs/models/dense_heads/clrernet_head.py#L429-L451)]
3. Form the lane segments with anchor x, y, length and prediction xs. [[code](https://github.com/hirotomusiker/CLRerNet/blob/main/libs/models/dense_heads/clrernet_head.py#L477-L505)]

### Angle definition

Traditionally (?), the angle of lane anchors are defined between 0 to pi/2. 
The following anchor definitions are useful. 
https://github.com/lucastabelini/LaneATT/blob/main/lib/models/laneatt.py#L40-L43

However, the lane angle output is not necessary for prediction.

**Comment by MickaMickaMicka** [comment]
Thank you so much, I was searching for the usage of theta during prediciton and couldn't identify it. 
If it's not used/needed during prediction that explains alot.
I thought the xs were offsets from the anchor-lanes.

**Comment by hirotomusiker** [comment]
Theta is not used in the postprocess but used in the `forward` function.
- The initial anchor xs calculation: [[code](https://github.com/hirotomusiker/CLRerNet/blob/main/libs/models/dense_heads/clrernet_head.py#L170)]
- Updating [[code](https://github.com/hirotomusiker/CLRerNet/blob/main/libs/models/dense_heads/clrernet_head.py#L222)]

So, if your onnx conversion fully includes `forward` function, you don't need to use theta. 

The output `xs` of `forward` are the sum of anchor xs and offsets. [[code](https://github.com/hirotomusiker/CLRerNet/blob/main/libs/models/dense_heads/clrernet_head.py#L230)]

**Comment by MickaMickaMicka** [comment]
Thank you, it looks like xs and ys are working well now. 

I still have some struggles with the start and end of a lane line.
Start index (from the 72 samples) is (1-anchor[0])*n_strips (clipped to 0 and n_strips) right ?
End index is start + length * n_strips?

However I am observing sometimes negative lengths values, is this expected? Should these samples be removed?
In addition I am facing many lanes that extend over the horizon, which is not the case with your repository code on the same images.

**Comment by MickaMickaMicka** [comment]
> Thank you, it looks like xs and ys are working well now.
> 
> I still have some struggles with the start and end of a lane line. Start index (from the 72 samples) is (1-anchor[0])*n_strips (clipped to 0 and n_strips) right ? End index is start + length * n_strips?
> 
> However I am observing sometimes negative lengths values, is this expected? Should these samples be removed? In addition I am facing many lanes that extend over the horizon, which is not the case with your repository code on the same images.

Finally it was a bug in my reading of the length dnn output values.
Now it's working as intended and the length values aren't negative.

We only find the CLRerNet to be very sensitive to image croppings (roi choices).
Is there some general advise how to set the roi for an arbitrary camera perspective?

**Comment by hirotomusiker** [comment]
Generally, to make your model robust to camera perspective, adding data augmentation for cropping works well.
CLRerNet is trained on three camera settings of CULane, so it could be sensitive to the camera pose.
